scm/build/cni_ds/entrypoint.sh

```diff
@@ -0,0 +1,16 @@
+#!/bin/sh -e
+
+echo "Copying plugins..."
+
+# Copy files into place and atomically move into final binary name
+for plugin in /cni/*
+do
+  plugin=$(basename "${plugin}")
+
+  echo "  - ${plugin}"
+  cp -fa /cni/"${plugin}" /host/cni/_"${plugin}"
+  mv -f  /host/cni/_"${plugin}" /host/cni/"${plugin}"
```

Comment:
Could you clarify why first `cp` to temp name, then rename with `mv`? 🤔

Reply:
Yes -- the comment on line 5 (not shown in the snipplet here, but a couple of lines up) was intended to address that :-)

**Long answer**

"mv" is an atomic operation; "cp" is not.

If someone else is attempting to execute the binary while we install it (which *is* possible in Kubernetes' declarative nature), then with "cp", there is a risk that they consume a partially-deployed binary.

Someone could have deployed multiple Kubernetes resources simultaneously; for example, someone could have `kubectl create`d the CNI daemonset, netwatcher, and svcwatcher at the same time -- in this case it is entirely possible that `svcwatcher` is already sitting and waiting to run the cni plugin, *while* the CNI plugin is being installed.

The risk is even higher after a node reboot, or in case of a CNI daemonset update (replacing the old CNI with a newer version). In this case, a whole set of apps might already be trying to (re-)start while the daemonset is restarting also (and will re-create the binaries). Again, in this case there'd be a risk that someone consumes an incomplete binary during a write-in-progress.

In both cases, this may result in unpredictable/undefined errors.

Copying the file to a temporary location and then doing a `mv` (which *is* an atomic operation), will prevent this. With `mv`, the consumer will either see no binary at all (in which case K8s will retry and wait for the plugin to be available), or a complete and sane binary - but never anything in between.


**Short answer**:

Other CNI plugins (eg. Multus <https://github.com/intel/multus-cni/blob/master/images/entrypoint.sh#L155>) do that, too ;-)

Reply:
My guess was the same, but I was not confident enough. I am okay with it.